Logout button

ThatJuanGuy opened the pull request:
## Description
This branch adds a working logout button to the website
## Rationale
Seeing as how it was possible to sign in, signing out seemed like the logical next step

## How to test
Login to the website. You should be logged in and your name and a logout button should appear where the login button used to be. Logout of the website, the login button should reappear and you should be logged out.

## Screenshots

|Before|After|
|--|--|
|Before Screenshot|After Screenshot|
|![LoggedOut](https://user-images.githubusercontent.com/16074716/92178559-6df2d000-ee08-11ea-8d32-d80bad42083d.JPG)|![LogoutIconOnlyArrow](https://user-images.githubusercontent.com/16074716/92334591-2a40d600-f055-11ea-894e-2303aa270daa.JPG)|


## Related tasks

Closes #329 #313

ThatJuanGuy commented:
Material UI only had the door and arrow logo as logout icons. Here's all the combinations of Logout and an icon. Personally, I like either Only Arrow and Door Then Logout. Of those two, I have a slight preference for Door then Logout. However, I wanted to get your opinions as to what yall think looks best. 

Only Door
![LogoutIconOnlyDoor](https://user-images.githubusercontent.com/16074716/92179004-6c75d780-ee09-11ea-80b0-31c983ba40e5.JPG)

Only Arrow
![LogoutIconOnlyArrow](https://user-images.githubusercontent.com/16074716/92178986-62ec6f80-ee09-11ea-91b4-7ea2874886c3.JPG)

Logout Then Arrow
![LogoutIconWordFirstArrow](https://user-images.githubusercontent.com/16074716/92179053-916a4a80-ee09-11ea-9e1c-404d25c95321.JPG)

Logout Then Door
![LogoutIconWordFirstDoor](https://user-images.githubusercontent.com/16074716/92179089-a21ac080-ee09-11ea-95a5-288b1747cfae.JPG)

Arrow Then Logout
![LogoutIconWordLastArrow](https://user-images.githubusercontent.com/16074716/92179104-aba42880-ee09-11ea-95b6-5db320772fce.JPG)

Door Then Logout
![LogoutIconWordLastDoor](https://user-images.githubusercontent.com/16074716/92179118-b19a0980-ee09-11ea-8032-5aa37e1e85ca.JPG)

firejake308 commented:
I prefer Only Arrow. For accessibility, we could add a label/alt text/tooltip to identify it as a logout button

ryan-conn commented on autoscheduler/frontend/src/components/NavBar.tsx:
```diff
@@ -23,6 +24,51 @@ const useStyles = makeStyles((theme) => ({
   const classes = useStyles(appTheme);
 
+  const [usersName, setUsersName] = React.useState('');
+  React.useEffect(() => {
```
Add a comment explaining that what this does (check if the user is logged in and sets name if so)

ryan-conn commented on autoscheduler/frontend/src/components/NavBar.tsx:
```diff
@@ -23,6 +24,51 @@ const useStyles = makeStyles((theme) => ({
+    }).catch(() => { });
+  }, []);
+
+  function LoginButton() {
+    return (
+      <Button
+        color="inherit"
+        onClick={(): void => {
+          window.open('/login/google-oauth2/', '_self');
+        }}
+      >
+        Login With Google
+      </Button>
+    );
+  }
+
+  function LogoutButton() {
+    return (
+      <Button
+        color="inherit"
+        aria-label="Logout"
+        title="Logout"
+        onClick={(): void => {
+          window.open('/sessions/logout', '_self');
+        }}
+      >
+        <ExitToAppOutlinedIcon />
+
+      </Button>
+    );
+  }
```
These don't have to be functions, you can do this:
```jsx
const userInformation = usersName ? (
  // component to render when logged in
) : (
  // component to render when logged out
);
```

ryan-conn commented on autoscheduler/frontend/src/components/NavBar.tsx:
```diff
@@ -44,15 +90,20 @@ const NavBar: React.SFC = () => {
-          >
-            Google Login
-          </Button>
+          {usersName === ''
+            ? (
+              <div>
+                {LoginButton()}
+              </div>
+            )
+            : (
+              <div>
+                <Typography variant="subtitle1">
+                  {usersName/* make font prettier */}
+                  {LogoutButton()}
+                </Typography>
+              </div>
+            )}
```
I like to avoid putting as little logic as possible into the component that gets rendered, so that you tell at a glance how the components are nested rather than looking at the business logic. You can do a couple things to help this:
- Move the ternary logic outside of the JSX
- Make the login/logout buttons include the typography, username, etc. instead of just the `Button` component

and for this part of the code, just put the variable that has the correct components to render:
```suggestion
          {userInformation}
```

ryan-conn commented on autoscheduler/user_sessions/views.py:
```diff
@@ -20,3 +23,19 @@ def set_last_term(request):
+    response = {'fullName': user.get_full_name()}
+    return Response(response)
+
+def logout_view(request):
+    """ Logs out the user and redirects to index"""
+    logout(request)
+    return redirect('/')
```
At the top of the file if you import logout like this:
```python
from django.contrib import auth
```
we can avoid having to call this `logout_view`, since none of our other views have `view` in the name:
```suggestion
def logout(request):
    """ Logs out the user and redirects to index"""
    auth.logout(request)
    return redirect('/')
```

ryan-conn reviewed (changes requested):
I agree with Adel's suggestion to just use the arrow icon, and also have a few ideas for making the NavBar more readable

gannonprudhomme commented:
Oh yeah when you log out it should give a toast that says like "You've successfully logged out" or something like that

ThatJuanGuy commented:
That’s a pretty good idea. I’ll add that.

On Tue, Sep 22, 2020 at 10:44 PM Gannon Prudhomme <notifications@github.com>
wrote:

>
>
> Oh yeah when you log out it should give a toast that says like "You've
> successfully logged out" or something like that
>
>
>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aggie-coding-club/Rev-Registration/pull/338#issuecomment-697116445>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD2UPXBKPJUWVRBWLTFCKETSHFVKNANCNFSM4QWIWVSQ>
> .
>
>
>

ryan-conn commented:
Also, as we talked about in our last meeting can you create a `LoginButton` component and move all of the login/logout functionality there instead of `NavBar`? It would be nice to isolate all the fetching, redirection, and state updates

firejake308 commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+import * as React from 'react';
+import { Button } from '@material-ui/core';
+import ExitToAppOutlinedIcon from '@material-ui/icons/ExitToAppOutlined';
```
Let's just rename it to `LogoutIcon` for our purposes. Saves a little space, and it also explains what it's doing better imo

ThatJuanGuy replied:
I had no idea I could rename this. Regardless it is done.

firejake308 commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  // used to logout the user
+  function logout(): void {
+    fetch('sessions/logout').then(
```
`sessions/logout` should definitely be a POST, not a GET

firejake308 commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+          window.location.reload();
+          return;
+        }
+        throw new Error(res.status.toString());
```
Why throw an error if you're going to catch it without handling it 2 lines later? Is that for future updates maybe?

ThatJuanGuy replied:
The catch is meant to stop the website from crashing if the API gives a bad response. It was in the getName UseEffect because when the response didn't have JSON it caused the website to crash. I made this one by copy pasting the getName one and then modifying it. Since I'm not actually processing any of the response other than checking `if(res.ok)` I can probably get rid of it here without any consequences.

firejake308 commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  // This checks for the logged in user's name and sets it if found.
+  // Otherwise, it throws an error and catches it so nothing breaks.
+  const [usersName, setUsersName] = React.useState('');
```
Although it technically doesn't matter, I like declaring all of the state at the top of the component because it makes it easier to find later

firejake308 commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+  }, []);
+
+  // Checks to make sure fetch call has finished before showing anything
+  if (loading) {
```
Could simplify to one line

ThatJuanGuy replied:
do you mean like `if (loading) return null;` ?

ThatJuanGuy replied:
If so, I am aware. I just left it the way I did because I like having the curly braces. I can change it no problem though if you want.

firejake308 commented on autoscheduler/frontend/src/components/App/App.tsx:
```diff
@@ -4,7 +4,7 @@ import { ThemeProvider } from '@material-ui/styles';
 import theme from '../../theme';
 import * as styles from './App.css';
-import NavBar from '../NavBar';
+import NavBar from '../NavBar/NavBar';
```
Interesting, it looks like the file should exist but linting says it doesn't. Weird.

firejake308 replied:
Oh wait, it's the caps. Linux is case-sensitive, so we'll need to rename the folder from `Navbar` to `NavBar`

ThatJuanGuy replied:
I thought I fixed this by renaming the folder before I pushed but apparently not. How do I make sure it gets renamed correctly? I'm thinking of deleting the folder, pushing that, adding it again with the correct casing, and pushing that but that seems pretty roundabout.

ryan-conn replied:
If you delete a file and add another identical one in the same commit, git will recognize that you renamed the file. That should be enough to fix this

firejake308 commented on autoscheduler/frontend/src/components/Navbar/NavBar.tsx:
```diff
@@ -53,15 +45,9 @@ const NavBar: React.SFC = () => {
-          >
-            {usersName}
-          </Button>
+          <Typography variant="subtitle1">
```
Why wrap it in a `<Typography>`?

ThatJuanGuy replied:
I believe I was doing that to control the text size. I could move the typography into the button component if it looks weird there.

firejake308 replied:
Ok, just move it into the `LogoutButton` component

firejake308 commented on autoscheduler/frontend/src/tests/ui/NavBar.test.tsx:
```diff
@@ -0,0 +1,153 @@
+import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
+// TODO ADD TEST FOR LOADING RETURN NULL
```
Reminder to address this TODO

firejake308 commented on autoscheduler/frontend/src/tests/ui/NavBar.test.tsx:
```diff
@@ -0,0 +1,153 @@
+    act(() => { fireEvent.click(logoutButton); });
+
+    // assert
+    expect(window.open).toHaveBeenCalledWith('/sessions/logout', '_self');
```
These tests look good to me. Nice work.

firejake308 commented on autoscheduler/user_sessions/views.py:
```diff
@@ -76,3 +77,9 @@ def get_full_name(request):
     response = {'fullName': user.get_full_name()}
     return Response(response)
+
+@api_view(['GET'])
```
Mentioned it earlier, but logout should be `POST` imo since it's not idempotent

ryan-conn commented on autoscheduler/frontend/src/components/Navbar/LoginButton.tsx:
```diff
@@ -0,0 +1,74 @@
+  return (
+    usersName ? (
+      <div>
+        {usersName}
```
Can you add `text-transform: uppercase` to this? Since the login text is uppercase, this should match it for consistency

ryan-conn replied:
Oh yeah also I realized when messing around with the text that the text and button aren't centered, also add `display: flex;` and `align-items: center;`

ryan-conn reviewed (changes requested):
Just a small change other than the ones Adel suggested, overall good work on this (especially the tests, they're pretty thorough and you didn't ask for help afaik despite not having much experience with them)

gannonprudhomme commented:
@ThatJuanGuy mind addressing these changes before the meeting? Would love to get this merged soon!

gannonprudhomme commented:
Could you edit this to make it pass the linting + testing? It's currently failing every frontend step

ThatJuanGuy commented:
> Could you edit this to make it pass the linting + testing? It's currently failing every frontend step. Been open for over a month so I'd like to get this + #366 merged before the website launch.

Just looked into it and it's because of the Navbar folder name. I deleted it before my last push and added it back with the correct casing (NavBar) but apparently the changes didn't take effect when I pushed. I'm just going to delete the folder, commit and push, add it back in a different commit and push again to see if that works.

ThatJuanGuy commented:
Still figuring out how to make the frontend tests for the refresh on logout and the button returning null if loading is true but website functionality wise it should be good to go.

firejake308 commented on autoscheduler/frontend/src/components/NavBar/LoginButton.tsx:
```diff
@@ -0,0 +1,91 @@
+          window.open('/login/google-oauth2/', '_self');
+        }}
+      >
+    Login With Google
```
Indent this line a little

firejake308 commented on autoscheduler/frontend/src/components/NavBar/NavBar.css:
```diff
@@ -0,0 +1,4 @@
+#NameAndButton {
+    display: flex;
+    align-items: center;
+}
```
Add new line

firejake308 commented on autoscheduler/frontend/src/components/NavBar/reloadPage.tsx:
```diff
@@ -0,0 +1,15 @@
+
+// This is written like this because the tests won't let me
+// replace the function with jest.fn() otherwise
+const reloadPageFunctions = {
```
This file could be simplified to just `export default function reloadPage...`
When you mock it, you'll have to do something like this:
```
jest.mock('../reloadPage', () => ({
  default: jest.fn(),
}));
```

firejake308 commented on autoscheduler/frontend/src/tests/ui/NavBar.test.tsx:
```diff
@@ -0,0 +1,166 @@
+    // as of time of creation of this test, mocking window.location.reload is impossible.
+    // Navigation is not supported by jsdom. Thus I am placing reload inside a function
+    // called reloadPage in reloadPage.tsx that holds window.location.reload. If the
+    // reloadPageFunctions.reloadPage was called, window.location.refresh is guranteed
```
```suggestion
    // reloadPageFunctions.reloadPage was called, window.location.refresh is guaranteed
```

firejake308 reviewed (changes requested):
So I can't test this in browser because I get this error:

![image](https://user-images.githubusercontent.com/10082177/96671179-e06f2f00-1326-11eb-8a86-f14eb0d34ea5.png)

But I had a few suggestions to clean up the code

ryan-conn commented on autoscheduler/frontend/src/components/NavBar/NavBar.css:
```diff
@@ -0,0 +1,4 @@
+#NameAndButton {
```
```suggestion
.name-and-button {
```
To be consistent with our other CSS rules, also since we're using classes instead of ids for like everything else on the site might as well conform to it

ryan-conn commented on autoscheduler/frontend/src/components/NavBar/LoginButton.tsx:
```diff
@@ -0,0 +1,91 @@
+          return res.json();
+        }
+
+        throw new Error(res.status.toString());
```
```suggestion
        throw new Error(String(res.status));
```
We should really be using `String()` instead of `.toString()` in places that use `.toString()` since it better fits typescript and can't throw errors, same for the other `.toString()` earlier in this file

ryan-conn commented on autoscheduler/frontend/src/components/NavBar/LoginButton.tsx:
```diff
@@ -0,0 +1,91 @@
+  // default set to defaultTestUserName instead of empty string becuase I thought
+  // an empty string might cause the refreshes on logout test to report false positives
+  // very easily depending on future changes.
+  const [usersName, setUsersName] = React.useState('defaultTestUserName');
```
```suggestion
  const [usersName, setUsersName] = React.useState<string>();
```
Use undefined as a placeholder value instead

ryan-conn reviewed (changes requested):
Couple of small changes and I'm ready to approve this
